Add st_b_lcd40_dsi1_mb1166 display shield support for stm32f769i_disco board

ssekar15 opened the pull request:
PR covers
- Add mipi ltdc display shield support for stm32f769i_disco board
- Replace direct hal api for ltdc block reset to zephyr reset api framework

ssekar15 committed:
dts: arm: st: Add STM32F7 DSI host driver

Add support for STM32F7 MIPI DSI host driver.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
dts: bindings: clock: Add pll-r for PLLDSICLK clock

Add bindings for pll-r used for DSI HOST peripheral byte clock.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
boards: stm32f769i_disco: add pll-r for PLLDSICLK clock

Add pll-r clock used for DSI HOST peripheral byte clock.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
boards: stm32f769i_disco: remove capacitive touch ft6202

Capacitive touch ft6202 is not part board, remove in board dts
and move to panel st_b_lcd40_dsi1_mb1166 shield.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
boards: shield: add support for stm32f769i_disco board

Add support for STMicroelectronics B-LCD40-DSI1 shield to
stm32f769i_disco board.

Default config for lvgl is adjusted to lower value to fix compilation
issue on low memory footprint like stm32f769i_disco.

region `IDT_LIST' overflowed by 121081 bytes
bss ' will not fit in region `RAM'
ld.bfd: region `RAM' overflowed by 1220664 bytes
collect2: error: ld returned 1 exit status

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

danieldegrasse commented on boards/shields/st_b_lcd40_dsi1_mb1166/boards/stm32f769i_disco.overlay:
If there are multiple ST boards that support the ST_B_LCD40_DSI1_MB1166 shield, then you might consider defining node labels for connectors (IE for the LTDC, MIPI DSI, and I2C connections). This has been done for some NXP supplied displays, see here: https://github.com/zephyrproject-rtos/zephyr/blob/59632d5baef3fa0f44394f8f8199b38f4d616ff4/boards/shields/rk055hdmipi4ma0/rk055hdmipi4ma0.overlay#L1. This avoids code duplication (for example, the display panel timings look the same as the ones here: https://github.com/zephyrproject-rtos/zephyr/blob/59632d5baef3fa0f44394f8f8199b38f4d616ff4/boards/shields/st_b_lcd40_dsi1_mb1166/boards/st_b_lcd40_dsi1_mb1166/stm32h747i_disco_stm32h747xx_m7.overlay#L1

It seems this is already being done for the GPIO nexus node, so expanding this to the peripherals might be useful

ssekar15 replied:
Touch and display controller are part of shield overlay. If I'm not wrong not all the STM chipset has LTDC, so the panel display timings in ltdc node shall be in board overlay.

danieldegrasse replied:
Sure, not all the STM SOCs will have an LTDC- this is the case with NXP's SOCs as well. In the case of NXP's displays, we sell multiple display modules that connect via a 40 pin FPC present on several of our evaluation kits to the module itself. The benefit here is that if we designed a new SOC with a DSI interface and added the same 40 pin connector, we could enable support for these displays by simply adding node definitions at the board level.

Beyond this, if a customer wanted to use this shield with their custom board, they would not need to modify Zephyr itself to add a new overlay (or duplicate the overlay definition for their board out of tree).

This is a core idea of a shield- not every evaluation kit supported by Zephyr has a SPI interface broken out to arduino headers, but those that do can use a large subset of all Zephyr shields *without board specific configuration*. Of course the scope of this shield would be more narrow (perhaps 2-3 boards), but I think the benefits to this approach still apply.

ajarmouni-st replied:
@ssekar15 see https://github.com/ajarmouni-st/zephyr_dev/tree/fix_mb1166_shield
@danieldegrasse is this more or less what you meant?

danieldegrasse replied:
Sorry for the delayed response- yeah, this looks right to me. Essentially just refactor the shields to avoid duplicating things in board-specific shield overlays

danieldegrasse commented on boards/st/stm32f769i_disco/stm32f769i_disco.dts:
```diff
@@ -135,19 +130,6 @@ arduino_serial: &usart6 {};
 	clock-frequency = <I2C_BITRATE_FAST>;
 };
 
-&i2c4 {
```
If I understand right, this display should be defined as a shield now to get the same functionality on the board? If so, we should put a note in the migration guide for 4.0 highlighting that a shield now needs to be set to use a display with this board for downstream users who are developing with this board and expect a display device to be defined.

ssekar15 replied:
there is no on-board display/touch and display support is added as new for this board using shield (updated in the board doc), I doubt this needs to me mentioned in migration guide. please correct me if my understanding is wrong.

ajarmouni-st replied:
@danieldegrasse This i2c node was added in https://github.com/zephyrproject-rtos/zephyr/pull/41519 to support an external touch controller.
In any case, the touch controller child node shouldn't be in the board's DTS, instead it should be in the overlay of each supported touchscreen shield.

ssekar15 committed:
boards: shield: add dsi-lcd display header mapping for stm32h747i_disco

Use common pin reference for STMicroelectronics B-LCD40-DSI1 shield from
board header mapping.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ajarmouni-st commented on boards/st/stm32f769i_disco/stm32f769i_disco.dts:
```diff
@@ -27,7 +27,7 @@
 		compatible = "zephyr,memory-region", "mmio-sram";
 		device_type = "memory";
-		reg = <0xc0000000 DT_SIZE_M(16)>;
+		reg = <0xc0000000 DT_SIZE_M(128)>;
```
incorrect, `DT_SIZE_M` is in MBytes

ssekar15 replied:
agree

ajarmouni-st commented on boards/shields/st_b_lcd40_dsi1_mb1166/boards/stm32f769i_disco.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_STM32_LTDC_RGB888=y
+CONFIG_HEAP_MEM_POOL_SIZE=65536
+# Initialize after LTDC and MIPI-DSI
+CONFIG_DISPLAY_OTM8009A_INIT_PRIORITY=87
```
This is specific to shield revisions that use the OTM8009A controller, so it should be in a dedicated `stm32f769i_disco.conf` in `boards/shields/st_b_lcd40_dsi1_mb1166/boards/st_b_lcd40_dsi1_mb1166`

ajarmouni-st commented on boards/st/stm32h747i_disco/stm32h747i_disco_stm32h747xx_m7.dts:
```diff
@@ -64,13 +64,36 @@
 	status = "okay";
 };
 
+touch_i2c: &i2c4 {
+	pinctrl-0 = <&i2c4_scl_pd12 &i2c4_sda_pd13>;
+	pinctrl-names = "default";
+	clock-frequency = <I2C_BITRATE_FAST>;
+};
+
```
Since this is not enabled, it shouldn't be in the base DTS file.

ssekar15 replied:
board schematics mentioned pd12 and pd13 i2c lines are used for connector CN1, so i2c node with pinctrl be part of board dts and overlays shall add i2c client device...

ajarmouni-st replied:
What I meant is that if you want to include a device node in the board's `.dts`, it should be enabled.

ssekar15 replied:
got it.. enabled in the dts, make sense to keep it on board dts because audio codec slave is on same i2c4 bus.

ajarmouni-st commented on boards/shields/st_b_lcd40_dsi1_mb1166/boards/stm32h747i_disco_stm32h747xx_m7.overlay:
```diff
@@ -61,15 +61,3 @@
 	hs-active-high;
 	de-active-high;
 };
-
-&i2c4 {
-	pinctrl-0 = <&i2c4_scl_pd12 &i2c4_sda_pd13>;
-	pinctrl-names = "default";
-	clock-frequency = <I2C_BITRATE_FAST>;
-	status = "okay";
-
-	ft5336: ft5336@38 {
-		compatible = "focaltech,ft5336";
-		reg = <0x38>;
-	};
-};
```
This should be kept here.

ssekar15 replied:
ft5336 touch controller is available in the panel, so added in the panel shield.

ajarmouni-st replied:
Yes, but the i2c device, that will be used with it, is board dependent.

ssekar15 commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -353,10 +357,6 @@ static int stm32_ltdc_init(const struct device *dev)
 	}
 #endif
 
-	/* reset LTDC peripheral */
-	__HAL_RCC_LTDC_FORCE_RESET();
-	__HAL_RCC_LTDC_RELEASE_RESET();
-
```
reset_line_toggle_dt is moved a few lines above.. I will correct it

ssekar15 commented on boards/shields/st_b_lcd40_dsi1_mb1166/Kconfig.defconfig:
IMHO I don't think lvgl setting default values are not relevant to the board. Either we can make it to low value to be compatible for all the boards or remove it..

ajarmouni-st replied:
some are board dependent & done to enhance performance.
`LV_Z_VBD_CUSTOM_SECTION` is enabled so that `boards/st/stm32h747i_disco/dc_ram.ld` creates a custom sections for LVGL buffers in SDRAM2.

ssekar15 replied:
ok, thanks for the info

ajarmouni-st commented on boards/shields/st_b_lcd40_dsi1_mb1166/Kconfig.defconfig:
```diff
@@ -8,19 +8,13 @@ if SHIELD_ST_B_LCD40_DSI1_MB1166 || SHIELD_ST_B_LCD40_DSI1_MB1166_A09
-
-config LV_Z_VBD_CUSTOM_SECTION
-	default y
+	default 10
```
`LV_Z_VDB_SIZE` default value is already `10` in `modules/lvgl/Kconfig.memory`.

ssekar15 replied:
when LV_Z_FULL_REFRESH  is selected then it turns out 100

ajarmouni-st reviewed (changes requested):
Looking great!! hopefully last remarques (haven't yet tested this on stm32f769i_disco)

ssekar15 commented:
> Looking great!! hopefully last remarques (haven't yet tested this on stm32f769i_disco)

Thanks for your time for the review.. 

I have tested samples/subsys/display/lvgl and samples/modules/lvgl/demos only on stm32f769i_disco

ajarmouni-st commented:
@ssekar15 stm32h747i_disco is looking good, will test stm32f769i_disco next.

ajarmouni-st commented:
> I have tested samples/subsys/display/lvgl and samples/modules/lvgl/demos only on stm32f769i_disco

@ssekar15 There is samples/drivers/display too.

ajarmouni-st commented:
> @ssekar15 stm32h747i_disco is looking good, will test stm32f769i_disco next.

@ssekar15 There seems to be a problem with `samples/drivers/display` on stm32f769i-disco, haven't looked into it yet.
![Screenshot_stm32f769i-disco](https://github.com/zephyrproject-rtos/zephyr/assets/128408655/d401e850-a330-440a-942b-2978a6db42fa)

ssekar15 commented:
> /

ok, what kind of problem? May I know how does correct one looks in other board?

ajarmouni-st commented:
> > /
> 
> ok, what kind of problem? May I know how does correct one looks in other board?

![screenshot_h747i](https://github.com/zephyrproject-rtos/zephyr/assets/128408655/03662d41-e95c-4ca6-9b66-ba98a6ccdd18)

ajarmouni-st commented:
> > /
> 
> ok, what kind of problem? May I know how does correct one looks in other board?

As a side note, you can run the sample on your PC (tested on Ubuntu laptop) using:
```
west build -p always -b native_posix/native/64 samples/drivers/display
west build -t run
```

ajarmouni-st reviewed (commented):
Adding the board to the different test cases of `st_b_lcd40_dsi1_mb1166` would also be needed.
```
diff --git a/samples/drivers/display/sample.yaml b/samples/drivers/display/sample.yaml
index 33d72e11c74..8e708140e68 100644
--- a/samples/drivers/display/sample.yaml
+++ b/samples/drivers/display/sample.yaml
@@ -132,7 +132,9 @@ tests:
       fixture: fixture_display
   sample.display.st_b_lcd40_dsi1_mb1166:
     filter: dt_compat_enabled("orisetech,otm8009a")
-    platform_allow: stm32h747i_disco/stm32h747xx/m7
+    platform_allow:
+      - stm32h747i_disco/stm32h747xx/m7
+      - stm32f769i_disco
     extra_args: SHIELD=st_b_lcd40_dsi1_mb1166
     tags:
       - display
diff --git a/samples/modules/lvgl/demos/sample.yaml b/samples/modules/lvgl/demos/sample.yaml
index 5a9d668d02d..eea1ed0b127 100644
--- a/samples/modules/lvgl/demos/sample.yaml
+++ b/samples/modules/lvgl/demos/sample.yaml
@@ -26,7 +26,9 @@ tests:
       - CONFIG_LV_Z_DEMO_WIDGETS=y
   sample.modules.lvgl.demos.st_b_lcd40_dsi1_mb1166:
     filter: dt_compat_enabled("orisetech,otm8009a")
-    platform_allow: stm32h747i_disco/stm32h747xx/m7
+    platform_allow:
+      - stm32h747i_disco/stm32h747xx/m7
+      - stm32f769i_disco
     extra_args: SHIELD=st_b_lcd40_dsi1_mb1166
     harness: console
     harness_config:
diff --git a/samples/subsys/display/lvgl/sample.yaml b/samples/subsys/display/lvgl/sample.yaml
index 94b26bc1087..c2715fe6c55 100644
--- a/samples/subsys/display/lvgl/sample.yaml
+++ b/samples/subsys/display/lvgl/sample.yaml
@@ -42,7 +42,9 @@ tests:
       - mimxrt1170_evk/mimxrt1176/cm7
   sample.subsys.display.lvgl.st_b_lcd40_dsi1_mb1166:
     filter: dt_compat_enabled("orisetech,otm8009a")
-    platform_allow: stm32h747i_disco/stm32h747xx/m7
+    platform_allow:
+      - stm32h747i_disco/stm32h747xx/m7
+      - stm32f769i_disco
     extra_args: SHIELD=st_b_lcd40_dsi1_mb1166
     harness: console
     harness_config:
```

ssekar15 committed:
samples: add disp_f769_disco to display samples with shield

Add disp_f769_disco board to platform-allow list for display sample tests
with st_b_lcd40_dsi1_mb1166 shield.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
samples: add disp_f769_disco to display samples with shield

Add disp_f769_disco board to platform-allow list for display sample tests
with st_b_lcd40_dsi1_mb1166 shield.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
samples: add disp_f769_disco to display samples with shield

Add disp_f769_disco board to platform-allow list for display sample tests
with st_b_lcd40_dsi1_mb1166 shield.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

erwango commented on dts/bindings/clock/st,stm32f7-pll-clock.yaml:
```diff
@@ -56,3 +56,9 @@ properties:
         PLL division factor for PLL48CK
         Valid range: 2 - 15
+
+  div-r:
```
Please update file top description field

erwango commented on boards/st/stm32h747i_disco/stm32h747i_disco_stm32h747xx_m7.dts:
```diff
@@ -64,13 +64,37 @@
 	};
+
+	dsi_lcd_qsh_030: connector_dsi_lcd {
+		compatible = "st,dsi-lcd-qsh-030";
```
Please add a dedicated yaml binding. See dts/bindings/gpio/arduino-header-r3.yaml as example

ajarmouni-st replied:
@erwango there is already one `dts/bindings/gpio/st,dsi-lcd-qsh-030.yaml`

erwango replied:
Ok, thanks, my bad.

erwango commented on boards/shields/st_b_lcd40_dsi1_mb1166/Kconfig.defconfig:
```diff
@@ -4,18 +4,14 @@
-# Double frame buffer maintained by lvgl.
 if LVGL
 
+if BOARD_STM32H747I_DISCO_STM32H747XX_M7
```
I'd rather add a `orsource "boards/*.defconfig"` then do this in a `boards/shields/st_b_lcd40_dsi1_mb1166/boards/stm32h747i_disco_stm32h747xx_m7.defconfig`

erwango commented:
Few comments from my part. Don't hesitate to dismiss my review once done, as will be OoO for few weeks.

ssekar15 commented:
> > @ssekar15 stm32h747i_disco is looking good, will test stm32f769i_disco next.
> 
> @ssekar15 There seems to be a problem with `samples/drivers/display` on stm32f769i-disco, haven't looked into it yet. !

@ajarmouni-st 
Related to display shift !

https://github.com/zephyrproject-rtos/zephyr/issues/32286

erwango reviewed (commented):
@ssekar15 One more comment that was left pending, but I'll need to review this entirely. Would you mind rebasing to solve merge conflicts ?

JarmouniA commented:
@ssekar15 Hello again from my personal account :)
Has the [shift issue](https://github.com/zephyrproject-rtos/zephyr/pull/74646#issuecomment-2202988246) on stm32f769i-disco been solved? I no longer have access to hardware to verify it.

ssekar15 commented:
> @ssekar15 Hello again from my personal account :) Has the [shift issue](https://github.com/zephyrproject-rtos/zephyr/pull/74646#issuecomment-2202988246) on stm32f769i-disco been solved? I no longer have access to hardware to verify it.

Unfortunately no.. I have tried with offset and different setting but didn't work

danieldegrasse commented:
> Nit: this line should be outside of if LVGL

@erwango is this your only outstanding comment? Just trying to understand the state of this PR, I don't want it to stall

danieldegrasse commented on boards/shields/st_b_lcd40_dsi1_mb1166/Kconfig.defconfig:
Just wanted to clarify the message of this commit:

```Enhanced lvgl config to improve performance for board stm32h747i_disco
impacts the compilation issue on low memory footprint stm32f769i_disco.

region `IDT_LIST' overflowed by 121081 bytes
bss ' will not fit in region `RAM'
ld.bfd: region `RAM' overflowed by 1220664 bytes
collect2: error: ld returned 1 exit status
```

Is this saying that this change prevents the `stm32f769i_disco` from compiling this shield? Or that this change fixes compilation for that target? We should strive to support both targets (and turn on performance options for the higher memory footprint one)

erwango commented:
@ssekar15 Please rebase

ssekar15 committed:
board: shield: fix compilation error on st_b_lcd40_dsi1_mb1166

Fixes: 8dd3aff158c9 ("board: shield: st_b_lcd40_dsi1_mb1166 A09 support")

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

zhang-wenchao commented:
I own this board and am trying to enable support. What is the current status of this PR? Does it require further maintenance?

zhang-wenchao commented:
I tested this PR and found that the screen displayed color cast and abnormal display orientation and the touch was not usable.

zhang-wenchao commented:
There is another PR that implements this feature.
